### DISCUSSION TITLE
Variable Eldoc line count

### DISCUSSION THREAD

**Discussion by Tiv0w** [emacs-lsp/lsp-mode#4270] [Ideas]
I've stumbled upon a minor inconvenience in regard to the way eldoc displays the `document/onHover` text.
Currently, there is a choice between showing 1 line and showing the entire text via `lsp-eldoc-render-all`. 

Metals (the language server for Scala) sometimes shows the signature on one line, and at other times it's a fully formatted Markdown text, having as first line `Expression type:`.

Here as one line:
![image](https://github.com/emacs-lsp/lsp-mode/assets/41124670/8fdaba6e-3de9-408c-8eed-46b5dbdca4b0)

Here as formatted text:
![image](https://github.com/emacs-lsp/lsp-mode/assets/41124670/c3506790-adc1-4412-a54c-31b6744d50d9)

I don't really like the minibuffer taking so much space to display hover, I prefer just having the signature, and when I need more documentation, calling `lsp-ui-doc-glance`.

So, I have tweaked the `lsp-clients-extract-signature-on-hover` to show a user-defined number of lines. Somewhat following https://github.com/emacs-lsp/lsp-mode/issues/2613, this could be a way to accommodate to the format imposed by the LSP (if it's non-"standard"), effectively letting users tweak "on-the-fly" the quantity of information given by eldoc.

```elisp
(use-package lsp-metals
  :after (scala-mode lsp-mode)
  :config
  (setq-local lsp-eldoc-render-lines-count 2))
```
![image](https://github.com/emacs-lsp/lsp-mode/assets/41124670/73423a25-498a-4b36-97f6-4e46fe02102b)

Would this change be appreciated in lsp-mode? (I can submit a PR.)

Many thanks to everyone involved in the project for this great suite of packages.

**Comment by yyoncho** [comment]
> So, I have tweaked the `lsp-clients-extract-signature-on-hover`

Technically,  `lsp-clients-extract-signature-on-hover` is extension point. You can write something like:

```
(cl-defmethod lsp-clients-extract-signature-on-hover (contents (_server-id (eql metals)))
 ;; the logic you like goes here
 )
```

Let me think for some time whether we want to have that configurable for on hover. We have such a configuration for signature help.

**Reply by Tiv0w** [reply]
Since yesterday I did exactly that, I changed `lsp-clients-extract-signature-on-hover` to filter some lines I don't want to show via regexp (namely the infamous `Expression type:` line from Metals).

But I don't really if it's that much necessary (nor wanted) to expose such behavior to the user of the `lsp-mode`, since the tinkerers will tinker, and it will probably add maintenance work.
